### ISSUE TITLE
SSL/TLS hostname verification 

### ISSUE THREAD

**Issue by Safihre** [nzbget/nzbget#332]
Fueled by @sanderjo we recently fixed the certificate verification part properly in SABnzbd.
But unfortunately many many usenet providers don't have proper certificates in place. 
The biggest problem is non matching hostnames which both SAB and nzbget don't check for, see my post on reddit here:
https://www.reddit.com/r/usenet/comments/5or1kt/using_ssl_in_sab_with_default_settings_or_nzbget/

I think if nzbget also supports the full verification, we can get more usenet providers to actually get proper certificates. Nowadays they don't have an excuse anymore not to do so.

**Comment by sanderjo** [comment]
So the problem is: the user has setup a NNTPS connection, it's working, so the user thinks it's safe and encrypted. Which it's not:

For example: 

NZBget will connect via SSL/TLS to news.tweakdsl.nl saying "Connection successful" 

However, news.tweakdsl.nl has an insecure SSL/TLS setup: it has a self-signed certifate, as reported by testssl.sh:

```
$ ./testssl.sh news.tweakdsl.nl:563
<snip>
 Issuer                       self-signed (NOT ok)
 Trust (hostname)             Ok via CN (works w/o SNI)
 Chain of trust               NOT ok (self signed)
```

A quick, poor man's test:

```
$ wget https://news.tweakdsl.nl:563
--2017-01-29 14:02:49--  https://news.tweakdsl.nl:563/
Resolving news.tweakdsl.nl (news.tweakdsl.nl)... 82.197.196.34
Connecting to news.tweakdsl.nl (news.tweakdsl.nl)|82.197.196.34|:563... connected.
ERROR: cannot verify news.tweakdsl.nl's certificate, issued by ‘emailAddress=info@tweakdsl.nl,CN=news.tweakdsl.nl,OU=System administration,O=TweakDSL,L=Almere,ST=FL,C=NL’:
  Self-signed certificate encountered.
To connect to news.tweakdsl.nl insecurely, use `--no-check-certificate'.
```

The default _SABnzbd_ setting will refuse this specific kind of connection. So maybe NZBget could warn/inform users too? This will make users more aware, and will push newsserver providers to really secure connections.

(Next post: less bad SSL/TLS connections, but still insecure)

**Comment by sanderjo** [comment]
Less bad SSL/TLS connections, but still insecure:

There is a real certificate, but it is not correct for the hostname:
```
$ wget https://news.thundernews.com:563
--2017-01-29 14:08:13--  https://news.thundernews.com:563/
Resolving news.thundernews.com (news.thundernews.com)... 2001:4de0:1::234, 2001:4de0:1::220, 81.171.92.220, ...
Connecting to news.thundernews.com (news.thundernews.com)|2001:4de0:1::234|:563... connected.
ERROR: no certificate subject alternative name matches
	requested host name ‘news.thundernews.com’.
To connect to news.thundernews.com insecurely, use `--no-check-certificate'.
```

Python will say: "CertificateError: hostname 'news.thundernews.com' doesn't match either of '*.sslusenet.com', 'sslusenet.com'". Currently SABnzbs' default setting is to accept these connection. Only reason: there are still very many newsservers will this kind of faulty SSL/TLS setup, most of them provider by Highwinds / sslusenet.com

In SABnzbd, you have to set to Strict to reject them. Hopefully we can make the strict checking the default setting in the near future.

**Comment by sanderjo** [comment]
And https://www.appelboor.com/newsservers/newsservers-with-SSL.html gives an overview of the SSL-correctness of newsservers.

**Comment by hugbug** [comment]
Certificate validation is a difficult thing for NZBGet. To perform validation a collection of trusted root certificates is needed. The big question is where to get that collection?

Interesting reading - [A note about SSL/TLS trusted certificate stores, and platforms (OpenSSL and GnuTLS)](https://www.happyassassin.net/2015/01/12/a-note-about-ssltls-trusted-certificate-stores-and-platforms/). A lot of confusion and this is just on Linux. Windows has it's own certificate store which of course isn't in format expected by OpenSSL. MacOS has it's own certificate store too.

An alternative could be to ship a certificate collection within NZBGet package. Curl project offers a [suitable file](https://curl.haxx.se/docs/caextract.html). Although wouldn't it be strange (or suspicious) that some program uses a non-system root certificate store? Well, this collection comes from Mozilla and Firefox does this exact thing (uses it's own certificate collection instead of system store).

----

So instead of trying to verify server certificate I could possibly implement a different approach to improve security and prevent MitM attacks. Which is a certificate fingerprint check. News servers configuration section in NZBGet will have a new option **Fingerprint** which must be set by user manually, something like:

> Server1.Host=news.tweakdsl.nl
> Server1.Fingerprint=EE:B2:B5:5D:58:09:33:C1:3D:63:2A:22:A3:1B:6D:D2

NZBGet will validate fingerprint and report an error and stop connecting if it doesn't match.

#### Advantages
- much more stronger security. No one can pretend to be the news server unless he has the original server certificate. If authorities create another certificate for the same host that certificate will have a different fingerprint and will be detected;
- it doesn't matter if certificate was issued to other site; what matters is if the user trusts to the certificate (solving sslusenet.com problem);
- self-signed certificates are as good as created by trusted issuers; what matters is if the user trusts to the certificate;
- no need to deal with certificate stores.

#### Disadvantages
- more work for users - instead of just activate one option "Validate certificate" they need to enter host fingerprint;
- if for some reason the server legitimately updates the certificate the users have to update fingerprint. Such situation is not distinguishable from an attack.

BTW I don't know how easy it is to implement fingerprint check, may be that's another challenge on it's own.

**Comment by hugbug** [comment]
## Command to obtain fingerprint

### SHA1:
```shell
openssl s_client -connect news.tweakdsl.nl:563 | openssl x509 -fingerprint -noout -sha1

...
SHA1 Fingerprint=98:8F:F9:40:07:4E:20:D5:4D:7F:9B:D2:83:20:EC:FB:DF:E5:C5:E7
```

### MD5:
```shell
openssl s_client -connect news.tweakdsl.nl:563 | openssl x509 -fingerprint -noout -md5

...
MD5 Fingerprint=EE:B2:B5:5D:58:09:33:C1:3D:63:2A:22:A3:1B:6D:D2
```

### SHA256:
```shell
openssl s_client -connect news.tweakdsl.nl:563 | openssl x509 -fingerprint -noout -sha256

...
SHA256 Fingerprint=CF:68:BF:95:4B:29:10:5E:F0:C9:C4:F9:5A:E4:A0:DF:19:2B:CC:9C:4E:E5:59:4A:B3:2E:30:6D:65:02:6E:A0
```

**Comment by Safihre** [comment]
The Disadvantages seem pretty big, especially since an MitM would look the same as a new certificate. It also doesn't sound user friendly..

I read the page about the Linux certificate store problems, which makes me think you could always try the default locations and if none of those are present, use a build-in Mozilla version?
Plus you could provide an (advanced) option for users to provide a path to their own storage, in case they don't trust yours.

**Comment by hugbug** [comment]
> especially since an MitM would look the same as a new certificate.

That can be a good thing actually. I assume law enforcement agencies can create valid certificates which would pass pure certificate validation without fingerprint check. They likely force one of the root certificate issuer (there are a LOT of them, just check the content of mozilla store file from the link above).

> It also doesn't sound user friendly.

That's the thing. Both solutions are not user friendly - think about Windows where there is no default cert store suitable. So either no validation for Windows users or shipping own cert store. Or a specialized solution for Windows to use the system cert store, that's tricky, I didn't find examples; digging into cpython may help but that's a lot of efforts.

**Comment by sanderjo** [comment]
> curl project offers a suitable file. 

Isn't that from Mozilla? https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/ says 

> Mozillaʼs CA Certificate Program governs inclusion of root certificates in Network Security Services (NSS), a set of open source libraries designed to support cross-platform development of security-enabled client and server applications. The NSS root certificate store is used in Mozilla products such as the Firefox browser, and is also used by other companies in a variety of products.

And https://hg.mozilla.org/mozilla-central/raw-file/tip/security/nss/lib/ckfw/builtins/certdata.txt seems to be the file you can include *in* your source code & binary? No external root certificate store needed? So not OS-dependent? And nzbget fullu self-contained with certificates ... ?

**Comment by hugbug** [comment]
> Isn't that from Mozilla?

Sure, I've mentioned that:

> this collection comes from Mozilla.

And the problems this creates:

> Although wouldn't it be strange (or suspicious) that some program uses a non-system root certificate store?

It's not a problem to obtain a list of root certificates valid at some time point (now): from Mozilla, from any Linux distribution, export from Mac OS keychain app, etc. The problems are:
- it's suspicious to provide a own root certificate list with your program;
- you have to maintain that list forever.

**Comment by sanderjo** [comment]
> Sure, I've mentioned that

Ah, sorry, apparently I hadn't fully understood you yet.

> it's suspicious to provide a own root certificate list with your program;

It's indeed all about trust and not trusting... :-)

Other programs have their non-OS root certificate too (Firefox, Chrome (AFAIK), and curl), and that is trusted by users. As NZBget is open source, anybody can inspect it and compile it themselves. I would certainly trust NZBget.

> you have to maintain that list forever.

I would say Mozilla / curl do the maintaining, and it would mean *including* the uptodate certdata.txt from Mozilla/curl in your program at the moment you compile it. 
AFAIK the root certificate is more static than nzbget releases.

FWIW: Probably redundant in this thread, but I found https://www.feistyduck.com/library/openssl-cookbook/online/ch-openssl.html  interesting.

> Building a Trust Store
> OpenSSL does not come with any trusted root certificates (also known as a trust store), so if you’re installing from scratch you’ll have to find them somewhere else. One possibility is to use the trust store built into your operating system. This choice is usually fine, but default trust stores may not always be up to date. A better choice—but one that involves more work—is to turn to Mozilla, which is putting a lot of effort into maintaining a robust trust store. For example, this is what I did for my assessment tool on SSL Labs.
> 
> Because it’s open source, Mozilla keeps the trust store in the source code repository:
> 
> https://hg.mozilla.org/mozilla-central/raw-file/tip/security/nss/lib/ckfw/builtins↩
> /certdata.txt
> Unfortunately, their certificate collection is in a proprietary format, which is not of much use to others as is. If you don’t mind getting the collection via a third party, the Curl project provides a regularly-updated conversion in Privacy-Enhanced Mail (PEM) format, which you can use directly:
> 
> http://curl.haxx.se/docs/caextract.html
> ... snip ...

Interesting stuff

**Comment by hugbug** [comment]
Made a test implementation for certificate validation.
Verification succeeds for astraweb and fails for thundernews. With disabled host name check thundernews works too. Seems to be all correct.

What I don't like is that the error message doesn't explain what the problem is:

> TLS handshake failed: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed

No mention about not matching host name. Do you become better error messages in SAB?

**Comment by sanderjo** [comment]
> Made a test implementation for certificate validation.

Wow!

If you get an error message in SAB, it's the message in the right hand column on https://www.appelboor.com/newsservers/newsservers-with-SSL.html

So, yes, 'hostname error', but still quite technical for normal users. I expect we will provide an URL in SABnzbd so users can get more information.

FYI: 
* the current Default setting in SABnzbd will accept OK OK OK and OK OK NOK. 
* If you set to Strict, SAB requires OK OK OK, and will print hostname errors in case of OK OK NOK.

Our goal / hope is to move to require to OK OK OK. Just lilke Chrome requires for websites.

**Comment by hugbug** [comment]
Can you give me a pointer, is the message like "CertificateError: hostname 'eu.newsdemon.com' doesn't match either of '*.sslusenet.com', 'sslusenet.com'" generated by SAB with complex analyzing of library error details or does the message come directly from pyhon?

**Comment by sanderjo** [comment]
Directly from python. I'll show you ... BRB 

EDIT 1: this is the python code that generates that list on appelboor.com: https://github.com/sanderjo/check-ssl-newsserver/blob/master/check_ssl_newsserver---to-csv.py

Ugly usage:

```
sander@Stream-13:~/git/check-ssl-newsserver$ python -c "import check_ssl_newsserver; print check_ssl_newsserver.check_NNTPS_server('news.thundernews.com',563,2)" 
No input, no valid host, or other error
('', '', 'ssl.CertificateError', "CertificateError: hostname 'news.thundernews.com' doesn't match either of '*.sslusenet.com', 'sslusenet.com'")
```

That's just the traceback the comes back from Python's connect with the context set to `context.verify_mode = ssl.CERT_REQUIRED` and `context.check_hostname = True`
So it's easy in Python.

But ... my guess: Python uses openssl, so the information must be in the info openssl returns ... ?

EDIT 2:

```
$ curl https://news.thundernews.com:563/
curl: (51) SSL: certificate subject name (*.sslusenet.com) does not match target host name 'news.thundernews.com'
```
So curl does know the reason.
Wild guess at the curl code:
```
    else if(!Curl_cert_hostcheck((const char *)peer_CN, hostname)) {
      failf(data, "SSL: certificate subject name '%s' does not match "
            "target host name '%s'", peer_CN, dispname);
```

... helpful?

**Comment by hugbug** [comment]
Curl has it's own hostname check because openssl didn't have host name check until 1.0.2.
I would prefer to use OpenSSL built in validation. It's easy to use, just set one parameter to activate the check. Only the error messages I don't like - the same message regardless of failure cause.

Python uses OpenSSL, it's possible they have their own check implementation as well. I'll take a look into python source.

**Comment by hugbug** [comment]
Guys, I'm wondering why do you offer "default"-option for certificate validation?

Isn't option "default" as insecure as "disabled"? Anyone can buy (hackers may also steal) a valid certificate for any valid web-site and then use it in MitM and SAB will not complain because the certificate is valid.

Wouldn't it be better to just have enabled/disabled options? Doesn't "default" give a false sense of better security compared to "disabled"?

**Comment by sanderjo** [comment]
> Guys, I'm wondering why do you offer "default"-option for certificate validation?

Because it's Python's default. And because "Strict" is a bridge too far as this moment looking at the state of newsservers' security:  see https://www.appelboor.com/newsservers/newsservers-with-SSL.html all those "OK OK NOK" servers :-(

> Isn't option "default" as insecure as "disabled"? Anyone can buy (hackers may also steal) a valid certificate for any valid web-site and then use it in MitM and SAB will not complain because the certificate is valid.

I don't think so. I would say you cannot buy a certificate for a domain that you have no control over. And if hackers steal your certifacte, yes, then there is security problem.

I think we need a phased approach with certificate checking for newsservers

1. Introduce awareness in usenet programs: Done for SABnzbd, hopefully for NZBget too
2. Introduce awareness at usenet providers / resellers. At this moment they are often in denial ("no, problem is in your client, not on our servers")
3. Introduce awareness at users. Offer security aware users the possibility to require full security
4. As soon as Highwinds / SSLusenet reseller get their setup correct, move to Strict checking as default

Remember: in the past, web browsers like Chrome also accepted non-secure connections, and in phases they went to more strict checking. If you're too strict too fast, you loose users.

**Comment by hugbug** [comment]
> I would say you cannot buy a certificate for a domain that you have no control over.

You can buy a certificate for your own domain and then use that certificate for attack. A hacker can also steal a certificate from ANY web-site and use it for attack.

> And if hackers steal your certifacte, yes, then there is security problem.

Not "your certificate" but "any certificate in the world". I think that's not hard for a specialist who can do MitM.

The "default" setting makes SAB happy with any certificate and that's not better than no check at all.

**Comment by Safihre** [comment]
@hugbug You are right. That's also what I wrote in the [Reddit post](https://www.reddit.com/r/usenet/comments/5or1kt/using_ssl_in_sab_with_default_settings_or_nzbget/). It's just an intermediate solution to raise awareness and to get rid of the self-signed ones.

We want Strict as the Default, but right now we would just scare away users, already the rejection of self-signed ones gave us quite some upset users that blamed SAB instead of the host.

@sanderjo and I just wanted to include you and NZBGet in the effort/discussion to make NNTPS actually secure, we're not saying that what we are doing is the best way :relaxed:

For the next SABnzbd release I am thinking of the following:
If a user has set Certificate-Validation to Default, we check if it's possible to connect with Strict. If this is not possible, we show a warning but still connect. If it connects successfully using Strict, we set it to Strict.
The message would be something like:
`Cannot fully validate identity of news.bla.com, the connection is not fully secure - See: https://sabnzbd.org/certificate-errors`
(we used this approach with an information URL also for other warnings so users don't flock to the forums, seems to be working well so far)

**EDIT:** 
We already heard from the guy behind Blocknews/Frugal that it is not that hard to add proper certificates as a Highwinds reseller, so all those in the list with `sslusenet.com` errors can do the same. They just need to be pushed a bit by users, and a message like the one above could push them!
